### PR TITLE
For #52: Adding EncounterBuilder

### PR DESCRIPTION
**Pull request by paulodamaso**
For #52: Adding EncounterBuilder

- added EncounterBuilder script with default generate method (takes random enemies from 'possible' list 'diffculty' times)
- left puzzles for putting the correct algorithm for enemies generation and to wire encounter builder in BattleSetup after #29 merge

### CODE REVIEW

**Comment by 0crat** [comment]
@silasreinagel[/z](https://www.0crat.com/u/silasreinagel) please, pay attention to this pull request

**Comment by 0crat** [comment]
Job [#113](https://github.com/enigmadragons/projectneon/issues/113) is now in scope, role is `REV`

**Comment by 0crat** [comment]
@silasreinagel[/z](https://www.0crat.com/u/silasreinagel) everybody who has role `REV` is banned at [#113](https://github.com/enigmadragons/projectneon/issues/113); I won't be able to assign anyone automatically; consider assigning someone manually (as in [§19](http://www.zerocracy.com/policy.html#19)), or invite more people (as in [§51](http://www.zerocracy.com/policy.html#51)), or remove the job from the scope (as in [§14](http://www.zerocracy.com/policy.html#14))

File: src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        /**
+         * @todo #52:30min Implement encounter generation after the encounter algorithm is defined, and then
+         *  return correct list of enemies in generate() method.
+         */
+        /**
+         * @todo #52:15min After #29 merge, wire encounter builder restult with BattleSetup so we can use the generated
+         *  enemies into the battle scene. BattleEnemiesSpawned  event must be triggered after they are added to
+         *  the scene.
+         */
+        if (difficulty < 0) difficulty = 1;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs]: This value changing is hidden from the designers. This violates the concept of "Don't Touch the Args".

Instead, you can limit the range of values in the inspector: https://www.reddit.com/r/Unity3D/comments/34mbrn/range_attribute_in_inspector_for_a_property/

File: src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        if (difficulty > 7) difficulty = 7;
+        List<Enemy> enemies = new List<Enemy>();
+        for (int i = 0; i < difficulty; i++)
+        {
+            enemies.Add(possible[Random.Range(0, possible.Count-1)]); ;
+        }
+        return possible;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs]: This should return the Selected Enemies, not the Possible enemies.

File: src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+         * @todo #52:15min After #29 merge, wire encounter builder restult with BattleSetup so we can use the generated
+         *  enemies into the battle scene. BattleEnemiesSpawned  event must be triggered after they are added to
+         *  the scene.
+         */
+        if (difficulty < 0) difficulty = 1;
+        if (difficulty > 7) difficulty = 7;
+        List<Enemy> enemies = new List<Enemy>();
+        for (int i = 0; i < difficulty; i++)
+        {
+            enemies.Add(possible[Random.Range(0, possible.Count-1)]); ;
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs]: Should be picking Enemies based on their Power Level, not just number of enemies. 

Modify Enemy to have a `Level` or `PowerLevel` or something and use that as the criteria.

File: src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+public class EncounterBuilder : MonoBehaviour
+{
+    [SerializeField]
+    private List<Enemy> possible;
+
+    [SerializeField]
+    private int difficulty;
+
+    public List<Enemy> generate()
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs]: C# Conventions - Public methods should be `PascalCased`.

File: src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    {
+        /**
+         * @todo #52:30min Implement encounter generation after the encounter algorithm is defined, and then
```
Comment (SilasReinagel) [inline src/ProjectNeon/Assets/Enemies/EncounterBuilder.cs]: Take at least a quick crack at making a basic algorithm here. Maybe just 5 minutes of work. 

If YOU were playing an RPG, what enemies would you expect to face?

Change this puzzle to Evolve Encounter Generation after playtesting. We want this Encounter Builder to work, even if it's really basic.

**Comment by paulodamaso** [comment]
@SilasReinagel Done, please take a look

**Comment by paulodamaso** [comment]
@SilasReinagel Well, I think that it is all right now: I've improved the method to create a enemy list with limit of 7 and the sum of their difficulties are less or equals the difficulty asked.

**Comment by paulodamaso** [comment]
@SilasReinagel Oh, just by the record, I'm used to create test for algorithms like this first implementing them. It is desired to do that this way in this project? If yes, how could we do that in the current project?

**Comment by SilasReinagel** [comment]
@paulodamaso This PR is fantastic! Great stuff.

Concerning tests, I created #118 and I will try to knock it out soon so that tests will be possible in this project.

**Comment by 0crat** [comment]
Job `gh:enigmadragons/projectneon#113` is not assigned, can't get performer

**Comment by 0crat** [comment]
The job [#113](https://github.com/enigmadragons/projectneon/issues/113) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 point(s) just awarded to @silasreinagel[/z](https://www.0crat.com/u/silasreinagel)

**Pull request merged**